src/core/qgsmaplayer.h

```diff
@@ -1001,6 +1002,22 @@ class CORE_EXPORT QgsMapLayer : public QObject
      */
     virtual void setDataSource( const QString &dataSource, const QString &baseName, const QString &provider, const QgsDataProvider::ProviderOptions &options, bool loadDefaultStyleFlag = false );
 
+    /**
+     * Updates the data source of the layer. The layer's renderer and legend will be preserved only
+     * if the geometry type of the new data source matches the current geometry type of the layer.
+     *
+     * Subclasses should override this method: default implementation does nothing.
+     *
+     * \param dataSource new layer data source
+     * \param baseName base name of the layer
+     * \param provider provider string
+     * \param options provider options
+     * \param flags provider read flags
+     * \see dataSourceChanged()
+     * \since QGIS 3.20
+     */
+    virtual void setDataSource( const QString &dataSource, const QString &baseName, const QString &provider, const QgsDataProvider::ProviderOptions &options, QgsDataProvider::ReadFlags flags );
```

Comment:
Can we make this method non-virtual and let it call an internal (private virtual) helper method instead?
Also can we deprecate the existing one?
The current approach where each subclass has the possibility to override 2 different setDataSource methods and we can't emit a centralized `dataSourceChanged` signal from the superclass isn't really API friendly.